lib/rules/color-contrast-matches.js

```diff
@@ -30,107 +42,80 @@ if (formElements.includes(nodeName)) {
-}
-
-if (nodeName === 'OPTION') {
-	return false;
-}
-
-if (
-	(nodeName === 'BUTTON' && node.disabled) ||
-	axe.commons.dom.findUpVirtual(virtualNode, 'button[disabled]')
-) {
-	return false;
-}
-
-if (
-	(nodeName === 'FIELDSET' && node.disabled) ||
-	axe.commons.dom.findUpVirtual(virtualNode, 'fieldset[disabled]')
-) {
-	return false;
+	// Match all form fields, regardless of if they have text
+	return true;
 }
 
 // check if the element is a label or label descendant for a disabled control
-const nodeParentLabel = axe.commons.dom.findUpVirtual(virtualNode, 'label');
-if (nodeName === 'LABEL' || nodeParentLabel) {
-	let relevantNode = node;
-	let relevantVirtualNode = virtualNode;
-
-	if (nodeParentLabel) {
-		relevantNode = nodeParentLabel;
-		// we need an input candidate from a parent to account for label children
-		relevantVirtualNode = axe.utils.getNodeFromTree(nodeParentLabel);
-	}
-	// explicit label of disabled input
-	const doc = axe.commons.dom.getRootNode(relevantNode);
-	let candidate =
-		relevantNode.htmlFor && doc.getElementById(relevantNode.htmlFor);
-	const candidateVirtualNode = axe.utils.getNodeFromTree(candidate);
-
-	if (
-		candidate &&
-		(candidate.disabled ||
-			candidate.getAttribute('aria-disabled') === 'true' ||
-			axe.commons.dom.findUpVirtual(
-				candidateVirtualNode,
-				'[aria-disabled="true"]'
-			))
-	) {
+const nodeParentLabel = dom.findUpVirtual(virtualNode, 'label');
+if (nodeName === 'label' || nodeParentLabel) {
+	let labelNode = nodeParentLabel || node;
+	let labelVirtual = nodeParentLabel
+		? axe.utils.getNodeFromTree(nodeParentLabel)
+		: virtualNode;
+
+	// explicit label of disabled control
+	const doc = dom.getRootNode(labelNode);
+	const explicitControl = doc.getElementById(labelNode.htmlFor || '');
+	const explicitControlVirtual =
+		explicitControl && axe.utils.getNodeFromTree(explicitControl);
+
+	if (explicitControlVirtual && forms.isDisabled(explicitControlVirtual)) {
 		return false;
 	}
 
-	candidate = axe.utils.querySelectorAll(
-		relevantVirtualNode,
+	// implicit label of disabled control
+	const query =
 		'input:not([type="hidden"]):not([type="image"])' +
-			':not([type="button"]):not([type="submit"]):not([type="reset"]), select, textarea'
-	);
-	if (candidate.length && candidate[0].actualNode.disabled) {
+		':not([type="button"]):not([type="submit"]):not([type="reset"]), select, textarea';
+	const implicitControl = axe.utils.querySelectorAll(labelVirtual, query)[0];
+
+	if (implicitControl && forms.isDisabled(implicitControl)) {
 		return false;
 	}
 }
 
-// label of disabled control associated w/ aria-labelledby
-if (node.getAttribute('id')) {
-	const id = axe.utils.escapeSelector(node.getAttribute('id'));
-	const doc = axe.commons.dom.getRootNode(node);
-	const candidate = doc.querySelector('[aria-labelledby~=' + id + ']');
-	if (candidate && candidate.disabled) {
-		return false;
+const ariaLabelledbyControls = [];
+let ancestorNode = virtualNode;
+while (ancestorNode) {
```

Comment:
This is going to very slow on large pages as it will check the entire parent tree for every node (include subnodes of nodes already checked). The result of this gathering should be cached on each of the nodes checked so the next subnode only has to look up 1 level to find the answer.

See https://github.com/dequelabs/axe-core/blob/develop/lib/core/utils/is-hidden.js for a similar approach.